File: Word Ladder/answer_TLE.cpp
// a BFS solution
// but it faces a TLE for large wordlist


class Solution {
public:
    // the boolean flag arry
    vector<bool>flag;
    // the queue for BFS
    // the first element is index of wordList
    // the second one is the ladder length from beginWord
    queue<pair<int,int>> q;
    
    // determine if two strings are modifiable via 1 letter
    bool isModifiable(string s1, string s2)
    {
        if(s1.size()!=s2.size())
            return false;
        int flag = 0;
        for(int i=0;i<s1.size();i++)
        {
            if(s1[i]!=s2[i]&&flag==0)
                flag++;
            else if(s1[i]!=s2[i]&&flag!=0)
                return false;
        }
        if(flag==1)
            return true;
        else
            return false;
    }
    
    // BFS searching
    int helper(string beginWord, string endWord, vector<string>& wordList){
        while(!q.empty())
        {
            int index = q.front().first;
            int num = q.front().second;
            q.pop();
            string word = index<0?beginWord:wordList[index];
            if(word == endWord)
                return num;
            if(index>=0)
                flag[index]=true;
            // cout<<num<<word<<"!!\t";
            for(int i=0;i<wordList.size();i++)
            {
                if(!flag[i]&&isModifiable(word, wordList[i]))
                {
                    // cout<<wordList[i]<<"\t";
                    q.push(make_pair(i,num+1));
                }
            }
            // cout<<endl;
        }
        return 0;
    }
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if(wordList.empty()||find(wordList.begin(),wordList.end(),endWord)>=wordList.end()||beginWord.size()!=endWord.size())
            return 0;
        // wordList.push_back(beginWord);
        flag = vector<bool>(wordList.size(), false);
        // push the beginWord;
        q.push(make_pair(-1, 1));
        return helper(beginWord, endWord, wordList);
    }
};

File: Word Ladder II/answer_TLE.cs
// interpreted from c++ version
// it is a TLE answer

public class Solution {
    // the boolean flag arry
    List<bool> flag;
    // the queue for BFS
    // the first queue is index of wordList
    // the second queue is the ladder length from beginWord
    
    Queue<int> indexQueue;
    Queue<List<string>> pathQueue;
    
    // key in allCombo is in the format with wild card, like h*t
    // the value of h*t is all possible string's index matching h*t
    Dictionary<string, List<int>> allCombo;

    // initialize allCombo
    public void Init(IList<string> wordList)
    {

        allCombo = new Dictionary<string, List<int>>();

        for(int i = 0;i<wordList.Count;i++)
        {
            string s = wordList[i];
            for(int j=0;j<s.Length;j++)
            {
                // we find all possibles with wildcard * matching wordList[i]
                StringBuilder wildCard = new StringBuilder(s);
                wildCard[j]='*';
                
		// beware of the initialization process
		if(allCombo.ContainsKey(wildCard.ToString()))
                    allCombo[wildCard.ToString()].Add(i);
                else  
                    allCombo.Add(wildCard.ToString(), new List<int>(){i});
            }
        }
    }
    IList<IList<string>> ret;
    int minRet = Int32.MaxValue;
    // BFS searcing
    public void Helper(string beginWord, string endWord, IList<string> wordList)
    {
        ret = new List<IList<string>>();
        while(indexQueue.Count>0)
        {
            int index = indexQueue.Dequeue();
            List<string> path = pathQueue.Dequeue();
            string word = index<0?beginWord:wordList[index];
            // Console.WriteLine(word);
            if(word == endWord)
            {
                // Console.WriteLine(path.Count);
                if(path.Count<=minRet)
                {
                    ret.Add(path);
                    minRet = path.Count;
                }
                else
                    return;
            }
            if(index>=0)
                flag[index]=true;
            
            // iterate all possble wildcard expressions
            // to find 1-lettter modifiable words from the word
            for(int i=0;i<word.Length;i++)
            {
                StringBuilder wildCard = new StringBuilder(word);
                wildCard[i] = '*';
                // cout<<wildCard<<"\t";
                if(!allCombo.ContainsKey(wildCard.ToString()))
                    continue;
                for(int j=0;j<allCombo[wildCard.ToString()].Count;j++)
                {
                    int possibleIndex = allCombo[wildCard.ToString()][j];
                    // cout<<wordList[possibleIndex]<<"\t"<<endl;
                    // Console.WriteLine(possibleIndex);
                    // Console.WriteLine(flag.Count);
                    if(!flag[possibleIndex])
                    {
                        // cout<<allCombo[wildCard][j]<<'\t';
                        path.Add(wordList[possibleIndex]);
                        indexQueue.Enqueue(possibleIndex);

			// in c++ versoin, the stack<T>::push is a copy constructor
			// here in C#, we need to initialize a new object
                        pathQueue.Enqueue(new List<string>(path));
                        path.RemoveAt(path.Count-1);
                    }
                }
            }
        }
        return ;
    }

    public IList<IList<string>> FindLadders(string beginWord, string endWord, IList<string> wordList) {
        if(wordList.Count==0)
            return new List<IList<string>>();
	
	// use LINQ to create a List, in which each element is false
        flag = Enumerable.Repeat(false, wordList.Count).ToList();
        // Console.WriteLine(wordList.Count);
        indexQueue = new Queue<int>();
        pathQueue = new Queue<List<string>>();
        indexQueue.Enqueue(-1);
        pathQueue.Enqueue(new List<string>(){beginWord});
        Init(wordList);
        Helper(beginWord, endWord, wordList);
        return ret;
    }
}
